contracts/Zauction.sol

```diff
@@ -121,6 +139,56 @@ contract ZAuction is Initializable, OwnableUpgradeable {
     );
   }
 
+  function setBuyPrice(uint256 amount, uint256 tokenId) external {
+    address owner = registrar.ownerOf(tokenId);
+    require(msg.sender == owner, "zAuction: only owner can set price");
+    priceInfo[tokenId] = Listing(amount, owner);
+  }
+
+  /// recovers buyer's signature based on seller's proposed data and, if bid data hash matches the message hash, transfers nft and payment
+  /// @param amount token amount of sale
+  /// @param tokenId token id we are transferring
+  /// @param startBlock block number at which acceptBid starts working
+  /// @param expireBlock block number at which acceptBid stops working
+  function buyNow(
+    uint256 amount,
+    uint256 tokenId,
+    uint256 startBlock,
+    uint256 expireBlock
+  ) external {
+    require(startBlock <= block.number, "zAuction: sale hasn't started");
+    require(expireBlock > block.number, "zAuction: sale expired");
+    require(amount == priceInfo[tokenId].price, "zAuction: wrong sale price");
```

Comment:
So I almost left a comment saying that this isn't necessary (passing in the `amount` as an argument to this function and then checking it vs the sale price) but then I realized this protects against a critical front running attack. Excellent execution on this, don't know if it was intentional or not since you're still getting the hang of smart contract dev.